File: apps/android/app/src/main/java/com/kelpie/browser/network/FeedbackStore.kt
package com.kelpie.browser.network

import android.content.Context
import org.json.JSONObject
import java.io.File
import java.time.Instant
import java.util.UUID

data class FeedbackRecord(
    val reportId: String,
    val storedAt: String,
    val payload: Map<String, Any?>,
)

object FeedbackStore {
    fun save(
        context: Context,
        body: Map<String, Any?>,
        platform: String,
        deviceId: String,
        deviceName: String,
    ): FeedbackRecord {
        val reportId = UUID.randomUUID().toString().lowercase()
        val storedAt = Instant.now().toString()
        val payload =
            body +
                mapOf(
                    "reportId" to reportId,
                    "storedAt" to storedAt,
                    "platform" to platform,
                    "deviceId" to deviceId,
                    "deviceName" to deviceName,
                )

        val directory = File(context.filesDir, "feedback").apply { mkdirs() }
        File(directory, "${storedAt.replace(":", "-")}-$reportId.json")
            .writeText(JSONObject(payload).toString(2))
        return FeedbackRecord(reportId = reportId, storedAt = storedAt, payload = payload)
    }
}

File: apps/macos/Kelpie/Network/FeedbackStore.swift
import Foundation

struct FeedbackRecord {
    let reportID: String
    let storedAt: String
    let payload: [String: Any]
}

enum FeedbackStore {
    static func save(
        payload body: [String: Any],
        platform: String,
        deviceID: String,
        deviceName: String
    ) throws -> FeedbackRecord {
        let reportID = UUID().uuidString.lowercased()
        let storedAt = ISO8601DateFormatter().string(from: Date())
        let payload = body.merging([
            "reportId": reportID,
            "storedAt": storedAt,
            "platform": platform,
            "deviceId": deviceID,
            "deviceName": deviceName
        ]) { _, new in new }

        let data = try JSONSerialization.data(withJSONObject: payload, options: [.prettyPrinted, .sortedKeys])
        try FileManager.default.createDirectory(at: feedbackDirectory(), withIntermediateDirectories: true)
        try data.write(to: feedbackDirectory().appendingPathComponent("\(storedAt.replacingOccurrences(of: ":", with: "-"))-\(reportID).json"))
        return FeedbackRecord(reportID: reportID, storedAt: storedAt, payload: payload)
    }

    private static func feedbackDirectory() -> URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? URL(fileURLWithPath: NSTemporaryDirectory(), isDirectory: true)
        return base
            .appendingPathComponent("Kelpie", isDirectory: true)
            .appendingPathComponent("feedback", isDirectory: true)
    }
}
